adafruit_minimqtt/adafruit_minimqtt.py

```diff
@@ -1046,14 +1046,23 @@ def _valid_qos(qos_level):
         else:
             raise MMQTTException("QoS must be an integer.")
 
-    def is_connected(self):
+    def _connected(self):
```

Comment:
Thanks, I see what you mean and your code is more elegant. I don't think we should change the behaviour though.

When I read the issue thread I had barely used the library, so I assumed that `is_connected` checked the current state of the connection to the MQTT broker. It does not. It checks to see whether `connect()` has been called before deciding whether code should be run and raises an exception if it should not. e. g. if we try to `publish()` before connecting.

The function behaves as designed. The problem is with the function's name.

`connect()` either connects successfully or raises an exception, so no need for the user to call `is_connected` in a while loop (which was what I thought I should do).

If the MQTT broker is down, the `is_connected` method will return `True` but any attempt to write to the broker will raise a `BrokenPipeError: 32` exception.

I propose that I close this pull request without merging any changes.

I could open a new pull request that renames the `is_connected` function to something like `_initial_connection_made` for a breaking change or one that edits the doc-string to make the function's purpose clear.

Reply:
Maybe I misunderstood your change, because I thought yours and mine would function identically.

 * is_connected returns a bool, like yours
 * _connected throws an exception if not is_connected, like yours

can you help me understand where you're seeing it as different? The return value of my _connected is None, not True, but as the return value is never used it didn't seem to matter as a difference.

Reply:
Hi, your change and mine do function the same. I'm not sure we should make the change at all. 

Here's why:

The `is_connected()` method only checks to see if there is a socket and if the variable `self._is_connected' is true.

`self._is_connected` is initialised `False` then set to `True` when `connect()` is called and a connection to the broker is successful. `self._is_connected` is set to `False` when `disconnect() is called.

The code calls 'is_connected()' at the beginning of methods like 'disconnect()`, 'publish()' and `will_set()`.

In `will_set`, for example:

```python
    if self._is_connected:
        raise MMQTTException("Last Will should only be called before connect().")
```

From reading through the code, I conclude that the `is_connected()` method has nothing to say about whether there is connection to the MQTT broker at the moment when it is called. The method is used by the author to decide whether it is appropriate to call methods based on whether a connection to whether `connect()` or `disconnect()` have previously been called.

Because the method is named `is_connected` I and others incorrectly assumed that the method will tell us whether there is a current, live connection to the MQTT server and think that the method should return `True` instead of raising an exception. We're wrong.

Therefore I think that changing `is_connected()` to return `True` instead of raising an exception will cause more problems for users than leaving it alone. The `ping()` method is the intended way to see if the broker connection is live or not.

I propose to cancel this pull request and open a new one to have a conversation around whether the `is_connected()` should be renamed to indicate that it is a private function, not intended to be called by from outside the module.

Reply:
Requested changes made, ready to go.